### ISSUE TITLE
Fix confusing naming of TaskOutputs.upToDateWhen

### ISSUE THREAD

**Issue by mschwaig** [gradle/gradle#4350]
I recently had to use `TaskOutputs.upToDateWhen` and I found the name of the method quite confusing so I would like to add an `outdatedUnless` method with the same semantics if you think that's a good idea.

For example:
`
task.outputs.upToDateWhen { someFileExists }`

to me this reads like "when some file exists the task is up to date" but this is not the documented behavior: https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskOutputs.html#upToDateWhen-groovy.lang.Closure-.

I think a method
`task.outputs.outdatedUnless { someFileExists }`

with the same semantics as `upToDateWhen` would be less confusing. This is especially true when there are many chained clauses since the reader tends to insert AND operators between those clauses in his head, which makes sense for `outdatedUnless ` but not for `upToDateWhen `. 

I think other people are also confused by the current naming:
https://discuss.gradle.org/t/outputs-uptodatewhen-confusion/17973

**Comment by mschwaig** [comment]
Another good name would be `onlyUpToDateWhen`.

**Comment by SUPERCILEX** [comment]
@mschwaig I'm confused, you said it reads like `when some file exists the task is up to date`, but that's exactly what the documentation states:
> If the closure returns false, the task outputs are considered **OUT-of-date**

How is that confusing? 😕

**Comment by grv87** [comment]
I see it confusing as soon as you have several `upToDateWhen`.
It is read as `if any one of them is true then the task is up-to-date` while in reality it is `if all of them are true, and also a bunch of other conditions`.

**Comment by lptr** [comment]
@grv87 I agree. I'm really interested in your use case, because I've been struggling for some time to find good use cases for this method. What conditions are you trying to use here with `upToDateWhen()`?

**Comment by grv87** [comment]
@lptr, if you mean `upToDateWhen` in general, then I used it in my Packer plugin. When Packer is expected to produce Amazon AMI I check whether such AMI (with specified name etc.) already exists. If yes and nothing else changed (sources) then task should be up-to-date.
In general, all tasks sending something to remote servers use `upToDateWhen`.

But it is one call for now, I don't have real use case for several calls yet.

**Comment by lptr** [comment]
Thanks @grv87.

**Comment by mschwaig** [comment]
We use ` upToDateWhen` in a few places, when there is some costly task we only need to run under certain conditions. Most of the time we check if some file or folder exists and only do the work if it does not.

A bug we had in that kind of logic is why I opened this issue. A Gradle task would run more often then necessary. Here is the relevant buggy code.
```
task expensiveOperationWithWorkingDirContents {
  outputs.upToDateWhen {
    !file(workingDir).exists()
  }
  [expensive operation]
}
```
This reads like the task would only run when workingDir exists, but in fact it always runs, because `upToDateWhen` creates a necessary condition for the task outputs to be up to date, not a sufficient condition as I would expect from the name.

@SUPERCILEXI do agree that the documentation is correct, but the name of the method is misleading.

I understand that this is very hard to fix since people are actively using `upToDateWhen`.

**Comment by lptr** [comment]
I'm not sure I agree with the method name being misleading. The intention here is for the code to (mostly) read as English:

> **task "expensiveOperationWithWorkingDirContents"** _'s_ **outputs** _are considered_ **up-to-date when** _the_ **file `workingDir` exists**

**Comment by lptr** [comment]
We want to do this.
